### ISSUE TITLE
Detailed Code Explanation - Sentiment Analysis Project

### ISSUE THREAD

**Issue by R2Vve** [R2Vve/NLP-projects#2]
# Line-by-Line Code Explanation

## 1. sentiment_analysis.py (Main Imports)

```python
import pandas as pd              # Data manipulation library
import numpy as np              # Numerical operations library
from sklearn.model_selection import train_test_split    # For splitting dataset
from sklearn.feature_extraction.text import TfidfVectorizer  # Text vectorization
from sklearn.linear_model import LogisticRegression    # Primary classifier
from sklearn.naive_bayes import MultinomialNB         # Bonus classifier
from sklearn.metrics import accuracy_score, classification_report  # Evaluation metrics
import nltk                     # Natural Language Toolkit
from nltk.corpus import stopwords    # For removing common words
from nltk.tokenize import word_tokenize  # For splitting text into words
import re                      # Regular expressions for text cleaning
import matplotlib.pyplot as plt  # For plotting
import seaborn as sns          # Enhanced plotting
```

Each import serves a specific purpose:
- pandas/numpy: Core data handling
- sklearn components: Machine learning functionality
- nltk: Text processing
- matplotlib/seaborn: Visualization

## 2. TextPreprocessor Class (text_preprocessor.py)

```python
class TextPreprocessor:
    def __init__(self):
        self.stop_words = set(stopwords.words('english'))  # Initialize English stopwords
```
- Constructor initializes stopwords set for efficient lookup

```python
    def preprocess_text(self, text):
        # Convert to lowercase
        text = text.lower()  # Standardize text case
        
        # Remove special characters and numbers
        text = re.sub(r'[^a-zA-Z\s]', '', text)  # Keep only letters and spaces
        
        # Tokenization
        tokens = word_tokenize(text)  # Split into individual words
        
        # Remove stopwords
        tokens = [token for token in tokens if token not in self.stop_words]
        
        # Join tokens back to text
        return ' '.join(tokens)  # Reconstruct cleaned text
```
Each step serves a purpose:
1. Lowercase conversion ensures consistency
2. Special character removal cleans the text
3. Tokenization breaks text into words
4. Stopword removal eliminates common words
5. Joining tokens recreates cleaned text

## 3. SentimentAnalyzer Class (sentiment_analyzer.py)

```python
class SentimentAnalyzer:
    def __init__(self):
        self.preprocessor = TextPreprocessor()  # Text cleaning component
        self.vectorizer = TfidfVectorizer(max_features=5000)  # Convert text to numbers
        self.classifier = LogisticRegression(max_iter=1000)   # Main classifier
        self.naive_bayes = MultinomialNB()  # Bonus classifier
```
Initialize all components needed for analysis

```python
    def prepare_data(self, df):
        # Preprocess all reviews
        df['processed_text'] = df['review'].apply(self.preprocessor.preprocess_text)
        
        # Split into training and testing sets
        X_train, X_test, y_train, y_test = train_test_split(
            df['processed_text'], 
            df['sentiment'], 
            test_size=0.2,   # 80% training, 20% testing
            random_state=42  # For reproducibility
        )
        
        # Vectorize the text
        X_train_vectorized = self.vectorizer.fit_transform(X_train)
        X_test_vectorized = self.vectorizer.transform(X_test)
        
        return X_train_vectorized, X_test_vectorized, y_train, y_test
```
Data preparation pipeline:
1. Clean all review texts
2. Split data into training/testing sets
3. Convert text to numerical features

## 4. WordFrequencyAnalyzer Class (word_frequency_analyzer.py)

```python
class WordFrequencyAnalyzer:
    def __init__(self):
        self.positive_words = Counter()  # Count positive words
        self.negative_words = Counter()  # Count negative words
```
Initialize counters for word frequency tracking

```python
    def analyze_word_frequencies(self, df, processed_text_column, sentiment_column):
        # Separate positive and negative reviews
        positive_texts = ' '.join(df[df[sentiment_column] == 1][processed_text_column])
        negative_texts = ' '.join(df[df[sentiment_column] == 0][processed_text_column])
        
        # Count frequencies
        self.positive_words = Counter(positive_texts.split())
        self.negative_words = Counter(negative_texts.split())
```
Word frequency analysis:
1. Separate reviews by sentiment
2. Join all text for each sentiment
3. Count word occurrences

## 5. ModelComparison Class (model_comparison.py)

```python
class ModelComparison:
    def __init__(self):
        self.models = {
            'Logistic Regression': LogisticRegression(max_iter=1000),
            'Naive Bayes': MultinomialNB()
        }
        self.results = {}  # Store evaluation results
```
Initialize models and results storage

```python
    def train_and_evaluate(self, X_train, X_test, y_train, y_test):
        for name, model in self.models.items():
            # Train model
            model.fit(X_train, y_train)
            
            # Make predictions
            y_pred = model.predict(X_test)
            
            # Store results
            self.results[name] = {
                'accuracy': accuracy_score(y_test, y_pred),
                'confusion_matrix': confusion_matrix(y_test, y_pred),
                'classification_report': classification_report(y_test, y_pred)
            }
```
Model evaluation process:
1. Train each model
2. Make predictions
3. Calculate and store metrics

## Key Concepts Used:

1. Text Processing:
   - Tokenization: Breaking text into words
   - Stopword removal: Eliminating common words
   - Case normalization: Converting to lowercase

2. Feature Engineering:
   - TF-IDF: Converting text to numerical features
   - Vectorization: Creating feature matrices

3. Machine Learning:
   - Model training: Fitting to training data
   - Prediction: Testing on new data
   - Evaluation: Calculating performance metrics

4. Visualization:
   - Word frequency plots
   - Confusion matrices
   - Performance comparisons

## Usage Instructions:

1. Data Preparation:
   ```python
   analyzer = SentimentAnalyzer()
   df = pd.read_csv('reviews.csv')
   X_train, X_test, y_train, y_test = analyzer.prepare_data(df)
   ```

2. Model Training:
   ```python
   comparison = ModelComparison()
   comparison.train_and_evaluate(X_train, X_test, y_train, y_test)
   ```

3. Visualization:
   ```python
   word_analyzer = WordFrequencyAnalyzer()
   word_analyzer.analyze_word_frequencies(df, 'processed_text', 'sentiment')
   word_analyzer.plot_word_frequencies()
   ```
